File: built-in plugins dsp/mb-eq-channel.cpp
/**
 * MB Channel EQ
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Console channel strip EQ
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_EQ_CHANNEL_H
#define MB_EQ_CHANNEL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEqChannel : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-eq-channel";
    static constexpr const char* PLUGIN_NAME    = "MB Channel EQ";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float low = 0f;  // range [-15, 15]
    float mid = 0f;  // range [-15, 15]
    float high = 0f;  // range [-15, 15]
    };

    MbEqChannel() = default;
    ~MbEqChannel() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.low = std::clamp(params.low, -15f, 15f);
        params.mid = std::clamp(params.mid, -15f, 15f);
        params.high = std::clamp(params.high, -15f, 15f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Channel EQ
        return input;
    }
};

#endif // MB_EQ_CHANNEL_H

File: built-in plugins dsp/mb-eq-channel.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbEqChannelPlugin: PluginDefinition = {
  id: "mb-eq-channel",
  slug: "mb-eq-channel",
  name: "MB Channel EQ",
  category: "effect",
  type: "eq",
  version: "1.0.0",
  description: "Console channel strip EQ",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "low",
      name: "Low",
      type: "float",
      defaultValue: 0,
      minValue: -15,
      maxValue: 15,
      automatable: true,
    },
    {
      id: "mid",
      name: "Mid",
      type: "float",
      defaultValue: 0,
      minValue: -15,
      maxValue: 15,
      automatable: true,
    },
    {
      id: "high",
      name: "High",
      type: "float",
      defaultValue: 0,
      minValue: -15,
      maxValue: 15,
      automatable: true,
    },
  ],
  defaultPreset: { low: 0, mid: 0, high: 0 },
};

export default MbEqChannelPlugin;
